napari/tests/test_napari.py

```diff
@@ -1,46 +1,70 @@
 import numpy as np
+import pytest
 import napari
 
+from napari.layers import LayerTypes
 
-def test_view_image(qtbot):
-    """Test adding image."""
 
-    np.random.seed(0)
-    data = np.random.random((10, 15))
+@pytest.mark.parametrize('layer_type, data, ndim, ndisplay', [
```

Comment:
This list of functions / data can actually be the same for both `test_napari.py` and `test_viewer.py` and these can be combined into one file to further reduce duplicate code. The only difference between them really is the use of 

```python
viewer = view_layer_type(layer_type, data, ndisplay)
```
and
```python
viewer = Viewer()
layer = viewer.add_layer_by_type(layer_type=layer_type, data=data)
```
i.e. how the viewer gets instantiated - as you can see lots of duplicate code!! Even the inside bits of the tests can be the same, and I left a bunch of comments in `test_viewer.py` on how it can be improved

Reply:
Assuming this parametrization is the only reason for `add_layer_by_type`'s existence, we can parametrize with:

```python
(Image, np.random.random((10, 15)), 2, None),
(Image, np.random.random((10, 15, 20)), 3, None),
...
viewer.add_layer(layer_type(data, ndisplay=ndisplay))
```

@sofroniewn given your comment about QtViewer and ViewerModel tests being sometimes identical, we can then write a utility function, in `napari/util/_testing.py`:

```python
def test_viewer(viewer, layer_type, data, **kwargs):
    viewer.add_layer(layer_type(data), **kwargs)
```

How does that sound?
and the

Reply:
While that change would work in some ways, it isn't quite identical - as say there is some divergence between say `viewer.add_image` and `viewer.add_layer(Image(...))`, with how the multichannel information is being handled for example, and the test should really be capturing the full `add_*` methods.

We could instead just parameterize by the actual functions themselves instead, no?:

```python
(viewer.add_image, np.random.random((10, 15)), 2, None),
(viewer.add_image, np.random.random((10, 15, 20)), 3, None),
(viewer.add_labels, np.random.randint(20, size=(10, 15, 20)), 3, None),
...
add_method(data)
```

and 

```python
(napari.view_image, np.random.random((10, 15)), 2, None),
(napari.view_image, np.random.random((10, 15, 20)), 3, None),
(napari.view_labels, np.random.randint(20, size=(10, 15, 20)), 3, None),
...
view_method(data, ndisplay=ndisplay)
```

In this implementation, this PR wouldn't need to add the enum or change our API in anyway and we'd still have full coverage on all our tests with much less duplicate code.

How does this approach sound @shanaxel42 @jni?

Reply:
@sofroniewn For the first section, you would need to instantiate a single viewer for all the methods, which I think makes the tests subtly different...

Reply:
hmm ok fair point, maybe at the end of the test we delete the added layer and check we have an empty viewer `assert len(viewer.layers) == 0`. That's probably a good piece of functionality to have tested regardless!!

Reply:
Things would be cleaner if they worked as @jni describes where `viewer.add_layer(Image(...)) == viewer.add_image()`, and actually even better if there was only the `add_layer` method. But since that's not true there needs to be some sort of mapping from a layer type to the correct add method. I'll try and work something out with the classes instead of the enum.

Reply:
how does one delete a layer?

Reply:
you should be able to treat `viewer.layers` like a normal list, so `pop` and `remove`